### PR TITLE
Move preparation of the ast to a new file

### PR DESCRIPTION
**Pull request by NSil**
Prepare the ast (desugar it, assign node ids) before doing anything complicated.

### CODE REVIEW

File: src/mar/build/mod.rs [hunk 1]
```diff
@@ -22,16 +22,19 @@ pub struct Builder<'a, 'b: 'a> {
...
 pub struct Error;
 
+
+
 ///////////////////////////////////////////////////////////////////////////
 // construct() -- the main entry point for building SMIR for a function
 
 pub fn construct(cx: &ExtCtxt, item: P<ast::Item>) -> Result<Mar, Error> {
-    let item = assign_node_ids(item);
+    let item = prepare_item(item).unwrap();
```
Comment (erickt) [inline src/mar/build/mod.rs]: Shouldn't this be `try!(prepare_item(item))`?

Comment (NSil) [inline src/mar/build/mod.rs]: It's a P, not an Option/Result

File: src/mar/build/prepare.rs [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+use syntax::codemap::{respan, Span};
+
+use mar::build::transition::ContainsTransition;
+
+pub fn prepare_item(item: P<Item>) -> P<Item> {
+    let mut expander = Expander::new();
+    let mut assigner = Assigner { next_node_id: 1 };
+
+    let expanded = fold::noop_fold_item_simple(item.unwrap(), &mut expander);
+    P(fold::noop_fold_item_simple(expanded, &mut assigner))
```
Comment (erickt) [inline src/mar/build/prepare.rs]: These could be written a little more simply with: `expander.fold_item_simple(item.unwrap())`

File: src/mar/build/prepare.rs [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+            attrs: fold::fold_attrs(attrs.into(), self).into(),
+        })
+    }
+
+}
+
+impl fold::Folder for Expander {
+
+    fn fold_expr(&mut self, e: P<Expr>) -> P<Expr> {
+
```
Comment (erickt) [inline src/mar/build/prepare.rs]: Nit: remove these extra newlines

**Pull request merged**